File: 14-TypeScript/Grey Multi-Tenant/packages/grey-dart/lib/src/domain/user_client.dart
/// User domain client for Grey SDK.
library;

import 'dart:async';

import 'package:grpc/grpc.dart';

import '../error/grey_error.dart';
import '../grpc/grey_channel.dart';
import '../grpc/types.dart';
import '../grpc/user_grpc_service.dart';

/// Client for user operations.
class UserClient {
  /// Creates a new user client.
  UserClient(this._channel, this._service);

  final GreyChannel _channel;
  final UserGrpcService _service;

  User? _cachedUser;

  /// Gets the current authenticated user.
  ///
  /// Returns cached data unless [forceRefresh] is true.
  ///
  /// Throws [GreyError] if not authenticated or on failure.
  Future<User> getUser({bool forceRefresh = false}) async {
    if (!_channel.isAuthenticated) {
      throw GreyError(
        code: GreyError.unauthorized,
        message: 'Not authenticated',
      );
    }

    // Return cached user if available and not forcing refresh
    if (!forceRefresh && _cachedUser != null) {
      return _cachedUser!;
    }

    try {
      final user = await _service.getUser(forceRefresh: forceRefresh);
      _cachedUser = user;
      return user;
    } on GrpcError catch (e) {
      throw GreyError.fromGrpcError(e);
    } catch (e) {
      throw GreyError.fromException(e);
    }
  }

  /// Clears the cached user data.
  void clearCache() {
    _cachedUser = null;
  }

  /// Gets the cached user without making a request.
  User? get cachedUser => _cachedUser;
}

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-scala/src/main/scala/com/grey/sdk/domain/UserClient.scala
package com.grey.sdk.domain

import com.grey.sdk.error.GreyError
import com.grey.sdk.grpc.{GreyChannel, User, UserGrpcService}
import java.util.concurrent.atomic.AtomicReference
import scala.concurrent.{ExecutionContext, Future}

/** Client for user operations.
  *
  * Provides user retrieval with optional caching.
  */
final class UserClient(
    channel: GreyChannel,
    service: UserGrpcService
)(using ec: ExecutionContext):

  private val cachedUser: AtomicReference[Option[User]] = AtomicReference(None)

  /** Gets the current authenticated user.
    *
    * Returns cached data unless forceRefresh is true.
    *
    * @param forceRefresh
    *   Whether to bypass cache and fetch fresh data
    * @return
    *   Future containing User on success
    * @throws GreyError
    *   if not authenticated or on failure
    */
  def getUser(forceRefresh: Boolean = false): Future[User] =
    if !channel.isAuthenticated then
      return Future.failed(GreyError.unauthorized())

    // Return cached if available and not forcing refresh
    if !forceRefresh then
      cachedUser.get() match
        case Some(user) => return Future.successful(user)
        case None       => // Continue to fetch

    service.getUser(forceRefresh).map { user =>
      cachedUser.set(Some(user))
      user
    }.recoverWith { case e => Future.failed(GreyError.fromThrowable(e)) }

  /** Gets user and returns Either instead of throwing. */
  def getUserSafe(forceRefresh: Boolean = false): Future[Either[GreyError, User]] =
    getUser(forceRefresh)
      .map(Right(_))
      .recover { case e: GreyError => Left(e) }

  /** Clears the cached user data. */
  def clearCache(): Unit = cachedUser.set(None)

  /** Gets the cached user without making a request. */
  def getCachedUser: Option[User] = cachedUser.get()

object UserClient:
  def apply(channel: GreyChannel, service: UserGrpcService)(using ExecutionContext): UserClient =
    new UserClient(channel, service)
